main/datastore_metric_upload.es6.js

```diff
@@ -0,0 +1,129 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+//
+// Compare local API data based on object graph JSON blobs to remote (Cloud
+// Datastore) data. Use count-of-entities as a proxy for "needs sync". Use
+// SET_MINUS to filter out existing data and upload missing data.
+//
+
+const fs = require('fs');
+const path = require('path');
+
+global.FOAM_FLAGS = {gcloud: true};
+require('foam2');
+require('../lib/datastore/datastore_container.es6.js');
+require('../lib/web_apis/release.es6.js');
+require('../lib/web_apis/release_interface_relationship.es6.js');
+require('../lib/web_apis/web_interface.es6.js');
+
+const credentials = JSON.parse(fs.readFileSync(
+    path.resolve(__dirname, '../.local/credentials.json')));
+
+const logger = global.logger =
+    foam.lookup('foam.nanos.log.ConsoleLogger').create();
+
+const datastoreCtx = global.datastoreCtx =
+    foam.lookup('org.chromium.apis.web.DatastoreContainer').create({
+      gcloudAuthEmail: credentials.client_email,
+      gcloudAuthPrivateKey: credentials.private_key,
+      gcloudProjectId: credentials.project_id,
+      logger: logger,
+    }).ctx;
+
+const MDAO = foam.lookup('foam.dao.MDAO');
+const Release = foam.lookup('org.chromium.apis.web.Release');
+const WebInterface = foam.lookup('org.chromium.apis.web.WebInterface');
+const ReleaseWebInterfaceJunction =
+    foam.lookup('org.chromium.apis.web.ReleaseWebInterfaceJunction');
+// Create context for computing metrics against local data:
+// - Local release, API, and relase/API DAOs for computing metrics,
+// - Remote browser metrics DAO pushing metrics to Datastore.
+const computeMetricsCtx = global.computeMetricsCtx =
+    foam.lookup('org.chromium.apis.web.DatastoreContainer').create({
+      releaseDAO: MDAO.create({of: Release}),
+      webInterfaceDAO: MDAO.create({of: WebInterface}),
+      releaseWebInterfaceJunctionDAO: MDAO.create({
+        of: ReleaseWebInterfaceJunction,
+      }),
+      // browserMetricsDAO: <default: Send to Datastore>,
+      gcloudAuthEmail: credentials.client_email,
+      gcloudAuthPrivateKey: credentials.private_key,
+      gcloudProjectId: credentials.project_id,
+      logger: logger,
+    }).ctx;
+
+// Add indexes to make relational lookups faster.
+computeMetricsCtx.releaseWebInterfaceJunctionDAO.addPropertyIndex(
+    ReleaseWebInterfaceJunction.SOURCE_ID);
+computeMetricsCtx.releaseWebInterfaceJunctionDAO.addPropertyIndex(
+    ReleaseWebInterfaceJunction.TARGET_ID);
+computeMetricsCtx.releaseWebInterfaceJunctionDAO.addPropertyIndex(
+    ReleaseWebInterfaceJunction.SOURCE_ID,
+    ReleaseWebInterfaceJunction.TARGET_ID);
+
+const AnonymousSink = foam.lookup('foam.dao.AnonymousSink');
+let promises = [];
+let computingMetrics = global.computingMetrics = false;
+function forwardDAOs(dao1, dao2) {
```

Comment:
Maybe source and destination instead of dao1 and dao2. Also maybe add some documentation.